### PR TITLE
Add custom error handling

### PR DESCRIPTION
**Pull request by rshipp**
### Description

Add custom error handling, and two new error views. One for generic server errors, and one for Pundit errors. Also adds the HTTP request ID to honeycomb spans (it's already in the Sentry tags), so we can find errors easier.

### Motivation and Context

Tasks: 
* https://app.asana.com/0/1174274412967132/1200663189674232/f
* https://app.asana.com/0/1174274412967132/1200523444029192/f

### Screenshots (if appropriate):
Screen Shot 2021-07-30 at 1 48 23 PM
Screen Shot 2021-07-30 at 1 49 52 PM

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] I have added neccessary labels
- [ ] My PR title follows the convention in the guide
- [ ] I have assigned myself to the PR
- [ ] I have requested reviews from team members
- [ ] I have created a task for reviewers on PM tool
- [ ] All new and existing tests passed.

### CODE REVIEW

File: app/controllers/home_controller.rb [hunk 1]
```diff
@@ -7,13 +7,17 @@ def welcome
...
     authorize :application, :index?
+
     # It's nice to get that page that shows all the routes in dev.
-    raise ActionController::RoutingError, "No route matches path /#{params[:other]}" if Rails.env.development?
-    redirect_to home_welcome_path
+    if Rails.env.development? && Rails.application.config.consider_all_requests_local
+      raise ActionController::RoutingError, "No route matches path /#{params[:other]}"
+    end
+
+    render plain: '404 Not Found', status: 404
```
Comment (rshipp) [inline app/controllers/home_controller.rb:21]: this render change is not really related, but I thought it made more sense than redirecting to the homepage. This also means things that previously redirected to the homepage inside LTI embeds, now 404 instead.

File: app/controllers/home_controller.rb [hunk 2]
```diff
@@ -7,13 +7,17 @@ def welcome
...
+  # in, return a 404 as expected.
   # This is so that bots can't enumerate paths and objects on our server.
   # Force an actual person to go to github to find that out!
   def missing_route
     authorize :application, :index?
+
     # It's nice to get that page that shows all the routes in dev.
-    raise ActionController::RoutingError, "No route matches path /#{params[:other]}" if Rails.env.development?
-    redirect_to home_welcome_path
+    if Rails.env.development? && Rails.application.config.consider_all_requests_local
```
Comment (rshipp) [inline app/controllers/home_controller.rb:17]: the change in the `if` logic here makes it so we can test error pages in development by setting `consider_all_requests_local = false`.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -83,6 +85,9 @@ def add_honeycomb_fields
     # will clobber all that and just overwrite them with the LtiLaunch values for those fields
     # in all spans in the trace.
     @lti_launch&.add_to_honeycomb_trace()
+
+    # Add the HTTP request ID.
+    Honeycomb.current_span&.add_field('request.id', request.request_id)
```
Comment (rshipp) [inline app/controllers/application_controller.rb:90]: https://ui.honeycomb.io/braven/datasets/development/result/bXQmhEoYj3M/trace/wvsv1tBxRqM

File: app/controllers/concerns/rescuable.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    @request = request
+    render 'errors/internal_server_error', layout: 'lti_canvas', status: 500
+  end
+
+  def capture_error(exception)
+    LOGGER.error e.backtrace.join("\n\t")
+      .sub("\n\t", ": #{e}#{e.class ? " (#{e.class})" : ''}\n\t")
+    Sentry.capture_exception(exception)
+    Honeycomb.client.send(:add_exception_data, Honeycomb.current_span, exception) if Honeycomb.current_span
```
Comment (rshipp) [inline app/controllers/concerns/rescuable.rb:37]: this seemed to be the safest/shortest way. if there's no current span, we won't get the exception in HC, but will still get it in Sentry. I don't think it's worth adding code to handle the no-span case, since we should have a span at this point in applicationcontroller.

File: app/views/errors/internal_server_error.html.erb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<h1>Oops!</h1>
+<p>Sorry, something's gone wrong.</p>
+<p>If you need help, please <%= link_to 'contact Braven Support', Rails.application.secrets.support_ticket_url %>,
+and include the following error details:</p>
+<pre>Path: <%= @request.path %>
```
Comment (rshipp) [inline app/views/errors/internal_server_error.html.erb:5]: i added the path just because people usually crop out the browser's url bar when they send screenshots, and having that info readily available would be way more helpful. the request id below will let us look things up in sentry and/or honeycomb if we need more info. I opted not to include any other information here since it's easy for us to look it up, and we don't want to disclose too much about server errors to users.

Comment (rshipp) [reply app/views/errors/internal_server_error.html.erb:5]: also note, I tested and this `path` does not include query params, so we won't leak `state` or anything here.

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (sadleb) [reply app/controllers/application_controller.rb:90]: Can you add to the comment about why we add to the current span, which is non standard in the codebase. E.g “We add it to the span so that it doesn’t get an ‘app’ prefix and matches the built-in request.id field that the Rails integration uses”

File: app/controllers/concerns/rescuable.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+  def handle_error_generic(exception)
+    capture_error(exception)
+
+    @request = request
+    render 'errors/internal_server_error', layout: 'lti_canvas', status: 500
+  end
+
+  def capture_error(exception)
+    LOGGER.error e.backtrace.join("\n\t")
```
Comment (sadleb) [inline app/controllers/concerns/rescuable.rb:34]: This line confuses me. What’s with the `LOGGER` constant? Where is `e` defined?

Comment (rshipp) [reply app/controllers/concerns/rescuable.rb:34]: I... am not sure how this worked. I copied it from the discord bot without thinking, tested it, saw the error in the logs, and moved on.

File: app/controllers/home_controller.rb (hunk 1 above)
Comment (sadleb) [reply app/controllers/home_controller.rb:21]: Seems fine. I can’t think of anything where redirecting to the homepage is behavior we rely on.

File: app/views/errors/internal_server_error.html.erb (hunk 5 above)
Comment (sadleb) [reply app/views/errors/internal_server_error.html.erb:5]: Makes sense

File: app/controllers/concerns/rescuable.rb (hunk 6 above)
Comment (rshipp) [reply app/controllers/concerns/rescuable.rb]: Fixed now
```
platformweb_1      |   ↳ app/models/user.rb:69:in `admin?'
platformweb_1      | /app/app/controllers/home_controller.rb:6:in `welcome': test message (StandardError)
platformweb_1      | 	/app/vendor/bundle/ruby/3.0.0/gems/actionpack-6.1.4/lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
```

Comment (rshipp) [reply app/controllers/concerns/rescuable.rb]: side note, this seems to bypass all the backtrace filtering, so lmk if you know of a better way to get backtraces into the logs.

Comment (sadleb) [reply app/controllers/concerns/rescuable.rb]: I think we’d have to dig through the Rails code and find how they do it using the backtrace silencer. Not worth it IMO

**Review by sadleb — APPROVED** [review]
Lgtm. I’m excited about this PR
